### PR TITLE
[IMP] Optimize cugraph storage sampling

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Optimize cugraph storage sampling

This PR optimizes cugraph sampling by 22x by removing the below serial loop: https://github.com/rapidsai/dgl/blob/3793623a5b89df2ad9c7462e1b5db742e7ef8666/python/dgl/contrib/cugraph/cugraph_storage.py#L129-L134 .  

This PR closes https://github.com/rapidsai/graph_dl/issues/8 .   

PR Time: 8.40285 vs 177.343 s (22x Delta) . 

The main advantage of the PR is to ensure that the sampling is bottlenecked by libcugraph op and not python host side code. 

**Mainline**
```python
Timer unit: 1e-06 s

Total time: 177.343 s
File: /home/nfs/vjawa/dgl/dgl/python/dgl/contrib/cugraph/cugraph_storage.py
Function: sample_neighbors at line 73

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    73                                               def sample_neighbors(self, seed_nodes, fanout, edge_dir='in', prob=None,
    74                                                                    exclude_edges=None, replace=False,
    75                                                                    output_device=None):
    76                                                   """
    77                                                   Return a DGLGraph which is a subgraph induced by sampling neighboring
    78                                                   edges ofthe given nodes.
    79                                                   See ``dgl.sampling.sample_neighbors`` for detailed semantics.
    80                                                   Parameters
    81                                                   ----------
    82                                                   seed_nodes : Tensor or dict[str, Tensor]
    83                                                       Node IDs to sample neighbors from.
    84                                                       This argument can take a single ID tensor or a dictionary of node
    85                                                       types and ID tensors. If a single tensor is given, the graph must
    86                                                       only have one type of nodes.
    87                                                   fanout : int or dict[etype, int]
    88                                                       The number of edges to be sampled for each node on each edge type.
    89                                                       This argument can take a single int or a dictionary of edge types
    90                                                       and ints. If a single int is given, DGL will sample this number of
    91                                                       edges for each node for every edge type.
    92                                                       If -1 is given for a single edge type, all the neighboring edges
    93                                                       with that edge type will be selected.
    94                                                   prob : str, optional
    95                                                       Feature name used as the (unnormalized) probabilities associated
    96                                                       with each neighboring edge of a node.  The feature must have only
    97                                                       one element for each edge.
    98                                                       The features must be non-negative floats, and the sum of the
    99                                                       features of inbound/outbound edges for every node must be positive
   100                                                       (though they don't have to sum up to one).  Otherwise, the result
   101                                                       will be undefined. If :attr:`prob` is not None, GPU sampling is
   102                                                       not supported.
   103                                                   exclude_edges: tensor or dict
   104                                                       Edge IDs to exclude during sampling neighbors for the seed nodes.
   105                                                       This argument can take a single ID tensor or a dictionary of edge
   106                                                       types and ID tensors. If a single tensor is given, the graph must
   107                                                       only have one type of nodes.
   108                                                   replace : bool, optional
   109                                                       If True, sample with replacement.
   110                                                   output_device : Framework-specific device context object, optional
   111                                                       The output device.  Default is the same as the input graph.
   112                                                   Returns
   113                                                   -------
   114                                                   DGLGraph
   115                                                       A sampled subgraph with the same nodes as the original graph, but
   116                                                       only the sampled neighboring edges.  The induced edge IDs will be
   117                                                       in ``edata[dgl.EID]``.
   118                                                   """
   119                                                   # change the seed_nodes from pytorch tensor to cudf series
   120         1         13.0     13.0      0.0          if torch.is_tensor(seed_nodes):
   121                                                       seed_nodes = cupy.asarray(seed_nodes)
   122                                                       seed_nodes = cudf.Series(seed_nodes)
   123                                           
   124         2    6246285.0 3123142.5      3.5          parents_nodes, children_nodes = self.graphstore.sample_neighbors(
   125         1          1.0      1.0      0.0              seed_nodes, fanout, edge_dir='in', prob=None, replace=False)
   126                                           
   127         1          3.0      3.0      0.0          num_edges = len(children_nodes)
   128         1         53.0     53.0      0.0          edge_ID_list = torch.zeros(num_edges)
   129      9778      13596.0      1.4      0.0          for i in range(num_edges):
   130      9777   34114739.0   3489.3     19.2              mask1 = self._edge_prop_df['_SRC_'] == int(parents_nodes[i])
   131      9777   39672566.0   4057.7     22.4              mask2 = self._edge_prop_df['_DST_'] == int(children_nodes[i])
   132     19554   97280880.0   4975.0     54.9              edge_ID_list[i] = torch.tensor(self._edge_prop_df[mask1 & mask2]
   133      9777      13170.0      1.3      0.0                                             ['_EDGE_ID_'].values_host)
   134                                                   # construct dgl graph, want to double check if children and parents
   135                                                   # are in the correct order
   136         1        966.0    966.0      0.0          sampled_graph = dgl.graph((children_nodes, parents_nodes))
   137                                                   # add '_ID'
   138         1          2.0      2.0      0.0          num_edges = len(children_nodes)
   139         1        225.0    225.0      0.0          sampled_graph.edata['_ID'] = edge_ID_list
   140                                                   # to device function move the dgl graph to desired devices
   141         1          1.0      1.0      0.0          if output_device is not None:
   142                                                       sampled_graph.to_device(output_device)
   143                                           
   144         1          0.0      0.0      0.0          return sampled_graph
```

**PR**
```
Total time: 8.40285 s
File: /home/nfs/vjawa/dgl/dgl/python/dgl/contrib/cugraph/cugraph_storage.py
Function: sample_neighbors at line 75

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    75                                               def sample_neighbors(
    76                                                   self,
    77                                                   seed_nodes,
    78                                                   fanout,
    79                                                   edge_dir="in",
    80                                                   prob=None,
    81                                                   exclude_edges=None,
    82                                                   replace=False,
    83                                                   output_device=None,
    84                                               ):
    85                                                   """
    86                                                   Return a DGLGraph which is a subgraph induced by sampling neighboring
    87                                                   edges ofthe given nodes.
    88                                                   See ``dgl.sampling.sample_neighbors`` for detailed semantics.
    89                                                   Parameters
    90                                                   ----------
    91                                                   seed_nodes : Tensor or dict[str, Tensor]
    92                                                       Node IDs to sample neighbors from.
    93                                                       This argument can take a single ID tensor or a dictionary of node
    94                                                       types and ID tensors. If a single tensor is given, the graph must
    95                                                       only have one type of nodes.
    96                                                   fanout : int or dict[etype, int]
    97                                                       The number of edges to be sampled for each node on each edge type.
    98                                                       This argument can take a single int or a dictionary of edge types
    99                                                       and ints. If a single int is given, DGL will sample this number of
   100                                                       edges for each node for every edge type.
   101                                                       If -1 is given for a single edge type, all the neighboring edges
   102                                                       with that edge type will be selected.
   103                                                   prob : str, optional
   104                                                       Feature name used as the (unnormalized) probabilities associated
   105                                                       with each neighboring edge of a node.  The feature must have only
   106                                                       one element for each edge.
   107                                                       The features must be non-negative floats, and the sum of the
   108                                                       features of inbound/outbound edges for every node must be positive
   109                                                       (though they don't have to sum up to one).  Otherwise, the result
   110                                                       will be undefined. If :attr:`prob` is not None, GPU sampling is
   111                                                       not supported.
   112                                                   exclude_edges: tensor or dict
   113                                                       Edge IDs to exclude during sampling neighbors for the seed nodes.
   114                                                       This argument can take a single ID tensor or a dictionary of edge
   115                                                       types and ID tensors. If a single tensor is given, the graph must
   116                                                       only have one type of nodes.
   117                                                   replace : bool, optional
   118                                                       If True, sample with replacement.
   119                                                   output_device : Framework-specific device context object, optional
   120                                                       The output device.  Default is the same as the input graph.
   121                                                   Returns
   122                                                   -------
   123                                                   DGLGraph
   124                                                       A sampled subgraph with the same nodes as the original graph, but
   125                                                       only the sampled neighboring edges.  The induced edge IDs will be
   126                                                       in ``edata[dgl.EID]``.
   127                                                   """
   128                                                   # change the seed_nodes from pytorch tensor to cudf series
   129         1         38.0     38.0      0.0          if torch.is_tensor(seed_nodes):
   130                                                       seed_nodes = to_dlpack(seed_nodes)
   131                                                       seed_nodes = cudf.from_dlpack(seed_nodes)
   132                                           
   133         2    8358552.0 4179276.0     99.5          parents_nodes, children_nodes = self.graphstore.sample_neighbors(
   134         1          1.0      1.0      0.0              seed_nodes, fanout, edge_dir="in", prob=None, replace=False
   135                                                   )
   136                                           
   137         1        354.0    354.0      0.0          sample_df = cudf.DataFrame({"_SRC_": parents_nodes, "_DST_": children_nodes})
   138         1          1.0      1.0      0.0          del parents_nodes, children_nodes
   139         2      40361.0  20180.5      0.5          edge_df = sample_df.merge(
   140         1        963.0    963.0      0.0              self._edge_prop_df[["_SRC_", "_DST_", "_EDGE_ID_"]], on=["_SRC_", "_DST_"]
   141                                                   )
   142         1         20.0     20.0      0.0          del sample_df
   143                                           
   144         1        381.0    381.0      0.0          edge_df = edge_df.astype(seed_nodes.dtype)
   145                                           
   146                                                   # construct dgl graph, want to double check if children and parents
   147                                                   # are in the correct order
   148                                           
   149         2        815.0    407.5      0.0          sampled_graph = dgl.graph(
   150         1          1.0      1.0      0.0              (
   151         1        486.0    486.0      0.0                  from_dlpack(edge_df["_DST_"].to_dlpack()),
   152         1        300.0    300.0      0.0                  from_dlpack(edge_df["_SRC_"].to_dlpack()),
   153                                                       )
   154                                                   )
   155                                           
   156         1        577.0    577.0      0.0          sampled_graph.edata["_ID"] = from_dlpack(edge_df["_EDGE_ID_"].to_dlpack())
   157                                           
   158                                                   # to device function move the dgl graph to desired devices
   159         1          1.0      1.0      0.0          if output_device is not None:
   160                                                       sampled_graph.to_device(output_device)
   161                                           
   162         1          0.0      0.0      0.0          return sampled_graph
​```

### CODE REVIEW

**Review by wangxiaoyunNV — APPROVED** [review]

**Comment by VibhuJawa** [comment]
@wangxiaoyunNV  , Are we good to merge this ?
